src/prerender.ts

```diff
@@ -0,0 +1,303 @@
+import path from "node:path";
+import os from "node:os";
+import { promises as fs } from "node:fs";
+
+import MagicString from "magic-string";
+import { parse } from "node-html-parser";
+
+import type { Plugin, UserConfig } from "vite";
+
+// Vite re-exports Rollup's type defs in newer versions,
+// merge into above type import when we bump the Vite devDep
+import type { InputOption, OutputAsset, OutputChunk } from "rollup";
+
+interface HeadElement {
+	type: string;
+	props: Record<string, string>;
+	children?: string;
+}
+
+interface Head {
+	lang: string;
+	title: string;
+	elements: Set<HeadElement>;
+}
+
+interface PrerenderedRoute {
+	url: string;
+	_discoveredBy?: PrerenderedRoute;
+}
+
+function enc(str: string) {
+	return str
+		.replace(/&/g, "&amp;")
+		.replace(/"/g, "&quot;")
+		.replace(/</g, "&lt;")
+		.replace(/>/g, "&gt;");
+}
+
+function serializeElement(
+	element: HeadElement | HeadElement[] | string,
+): string {
+	if (element == null) return "";
+	if (typeof element !== "object") return String(element);
+	if (Array.isArray(element)) return element.map(serializeElement).join("");
+	const type = element.type;
+	let s = `<${type}`;
+	const props = element.props || {};
+	let children = element.children;
+	for (const prop of Object.keys(props)) {
+		const value = props[prop];
+		// Filter out empty values:
+		if (value == null) continue;
+		if (prop === "children" || prop === "textContent") children = value;
+		else s += ` ${prop}="${enc(value)}"`;
+	}
+	s += ">";
+	if (!/link|meta|base/.test(type)) {
+		if (children) s += serializeElement(children);
+		s += `</${type}>`;
+	}
+	return s;
+}
+
+interface PrerenderPluginOptions {
+	prerenderScript?: string;
+	additionalPrerenderRoutes?: string[];
+}
+
+export function PrerenderPlugin({
+	prerenderScript,
+	additionalPrerenderRoutes,
+}: PrerenderPluginOptions = {}): Plugin {
+	const preloadHelperId = "vite/preload-helper";
+	let viteConfig: UserConfig = {};
+
+	additionalPrerenderRoutes ||= [];
+
+	/**
+	 * Retrieves the last, non-external script from the entry HTML document to use as
+	 * an additional input for Rollup. Hopefully this contains the user's `prerender()`
+	 * function.
+	 */
+	const guessPrerenderScriptFromHTML = async (input: InputOption) => {
```

Comment:
Unfortunately I'm not sure if there's any better method? Without running Babel on each script to handle JSX, TS, etc. or a hacky regex, I think we're left to guessing?

My kingdom for a version of [es-module-lexer](https://github.com/guybedford/es-module-lexer) that supported JSX & TS

Reply:
There is https://github.com/fz6m/rs-module-lexer which is based on swc. It supports JSX + TS.

Reply:
OMG I love you. I spent a whole bunch of time searching for something like that a few months back, was starting to patch `es-module-lexer` as I really needed that for some other projects.

Well, time to hand over my kingdom I guess...